File: desktop/src/protocol.ts
export const VERSION = 1 as const;

export const LIMITS = Object.freeze({
  maxFrameBytes: 65_536,
  maxOutstanding: 8,
  maxPageSize: 128,
  maxSceneItems: 128,
  progressIntervalMs: 100,
});

export const OPERATIONS = [
  "capabilities",
  "snapshot.load",
  "snapshot.dispose",
  "query",
  "scene",
  "details",
  "recipe.validate",
  "export",
] as const;

export type Operation = (typeof OPERATIONS)[number];
export type Capabilities = {
  backend: "fake";
  operations: readonly [
    "capabilities",
    "snapshot.load",
    "snapshot.dispose",
    "query",
    "scene",
    "details",
    "recipe.validate",
    "export",
  ];
  extensions: readonly [];
  limits: typeof LIMITS;
};

export const CAPABILITIES: Capabilities = Object.freeze({
  backend: "fake",
  operations: Object.freeze(OPERATIONS),
  extensions: Object.freeze([]) as readonly [],
  limits: LIMITS,
});

export type Hello = {
  tag: "hello";
  versions: number[];
  extensions: string[];
};

type RequestBase = {
  tag: "request";
  version: typeof VERSION;
  requestId: string;
};

export type Request = RequestBase &
  (
    | {
        operation: "capabilities";
        snapshotId: null;
        args: Record<string, never>;
      }
    | {
        operation: "snapshot.load";
        snapshotId: null;
        args: { source: "fixture:tiny"; delayMs: number };
      }
    | {
        operation: "snapshot.dispose";
        snapshotId: string;
        args: Record<string, never>;
      }
    | {
        operation: "query";
        snapshotId: string;
        args: { text: string; pageSize: number; cursor: string | null };
      }
    | {
        operation: "scene";
        snapshotId: string;
        args: { maxItems: number };
      }
    | {
        operation: "details";
        snapshotId: string;
        args: { objectId: string; pageSize: number; cursor: string | null };
      }
    | {
        operation: "recipe.validate";
        snapshotId: null;
        args: { recipe: { schemaVersion: typeof VERSION; query: string } };
      }
    | {
        operation: "export";
        snapshotId: string;
        args: { format: "svg"; maxBytes: number };
      }
  );

export type Cancel = {
  tag: "cancel";
  version: typeof VERSION;
  requestId: string;
};
export type Shutdown = { tag: "shutdown"; version: typeof VERSION };
export type Inbound = Hello | Request | Cancel | Shutdown;

export type Item = { objectId: string; address: string; size: string };
export type Page = {
  tag: "page";
  items: Item[];
  nextCursor: string | null;
  truncated: boolean;
};
export type Scene = {
  tag: "scene";
  schemaVersion: typeof VERSION;
  items: Item[];
  truncated: boolean;
};
export type Result =
  | { tag: "capabilities"; value: Capabilities }
  | { tag: "snapshot"; objectCount: string }
  | { tag: "disposed" }
  | Page
  | Scene
  | { tag: "recipe"; schemaVersion: typeof VERSION; valid: true }
  | {
      tag: "export";
      format: "svg";
      artifactId: "fixture:svg";
      byteLength: string;
    };

export type Success = {
  tag: "success";
  version: typeof VERSION;
  requestId: string;
  snapshotId: string | null;
  result: Result;
};

export const ERROR_CODES = [
  "Busy",
  "Cancelled",
  "SnapshotNotFound",
  "InvalidRequest",
  "NotImplemented",
  "InternalError",
  "StaleSnapshot",
  "WorkerExited",
  "Timeout",
  "ProtocolError",
] as const;
export type ErrorCode = (typeof ERROR_CODES)[number];
export type ErrorResponse = {
  tag: "error";
  version: typeof VERSION;
  requestId: string;
  snapshotId: string | null;
  error: { code: ErrorCode; message: string; retryable: boolean };
};
export type Progress = {
  tag: "progress";
  version: typeof VERSION;
  requestId: string;
  snapshotId: string | null;
  phase: "working";
  completed: number;
  total: 100;
};
export const FATAL_CODES = [
  "UnsupportedVersion",
  "UnsupportedExtension",
  "InvalidFrame",
  "ProtocolViolation",
  "TransportTimeout",
  "InternalError",
] as const;
export type Fatal = {
  tag: "fatal";
  code: (typeof FATAL_CODES)[number];
  message: string;
};
export type Ready = {
  tag: "ready";
  version: typeof VERSION;
  capabilities: Capabilities;
};
export type Bye = { tag: "bye"; version: typeof VERSION };
export type Outbound = Ready | Fatal | Success | ErrorResponse | Progress | Bye;

export class ProtocolError extends Error {
  constructor(
    public readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = "ProtocolError";
  }
}

function invalid(): never {
  throw new ProtocolError("ProtocolViolation", "Invalid protocol message.");
}

type ObjectValue = Record<string, unknown>;

function object(value: unknown): ObjectValue {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return invalid();
  }
  const prototype: unknown = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) return invalid();
  return value as ObjectValue;
}

function fields(value: unknown, names: readonly string[]): ObjectValue {
  const result = object(value);
  const keys = Object.keys(result);
  if (keys.length !== names.length || !names.every((name) => Object.hasOwn(result, name))) {
    return invalid();
  }
  return result;
}

function equal(value: unknown, expected: unknown): void {
  if (value !== expected) invalid();
}

export function isValidUnicode(value: string): boolean {
  for (let i = 0; i < value.length; i++) {
    const unit = value.charCodeAt(i);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(++i);
      if (!(next >= 0xdc00 && next <= 0xdfff)) return false;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      return false;
    }
  }
  return true;
}

function text(value: unknown, min: number, max: number): string {
  if (typeof value !== "string") return invalid();
  let length = 0;
  for (const character of value) {
    if (++length > max || !isValidUnicode(character)) return invalid();
  }
  if (length < min) return invalid();
  return value;
}

function integer(value: unknown, min: number, max: number): number {
  if (
    typeof value !== "number" ||
    !Number.isSafeInteger(value) ||
    Object.is(value, -0) ||
    value < min ||
    value > max
  ) {
    return invalid();
  }
  return value;
}

function array(value: unknown, min: number, max: number): unknown[] {
  if (!Array.isArray(value) || value.length < min || value.length > max) {
    return invalid();
  }
  return value as unknown[];
}

function boolean(value: unknown): void {
  if (typeof value !== "boolean") invalid();
}

function member(value: unknown, values: readonly string[]): void {
  if (typeof value !== "string" || !values.includes(value)) invalid();
}

function uint64(value: unknown, nonzero = false): void {
  const decimal = text(value, 1, 20);
  // Unlike $, this end assertion cannot match before a final line terminator.
  if (
    !/^(?:0|[1-9][0-9]*)(?![\s\S])/.test(decimal) ||
    (nonzero && decimal === "0") ||
    (decimal.length === 20 && decimal > "18446744073709551615")
  ) {
    invalid();
  }
}

function snapshot(value: unknown): void {
  const uuid = text(value, 36, 36);
  if (
    !/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}(?![\s\S])/.test(uuid) ||
    uuid === "00000000-0000-0000-0000-000000000000"
  ) {
    invalid();
  }
}

function nullableSnapshot(value: unknown): void {
  if (value !== null) snapshot(value);
}

function cursor(value: unknown): void {
  if (value !== null) uint64(value);
}

function capabilities(value: unknown): void {
  const caps = fields(value, ["backend", "operations", "extensions", "limits"]);
  equal(caps.backend, "fake");
  const operations = array(caps.operations, OPERATIONS.length, OPERATIONS.length);
  for (let i = 0; i < OPERATIONS.length; i++) {
    equal(operations[i], OPERATIONS[i]);
  }
  array(caps.extensions, 0, 0);
  const limits = fields(caps.limits, Object.keys(LIMITS));
  for (const key of Object.keys(LIMITS) as (keyof typeof LIMITS)[]) {
    equal(limits[key], LIMITS[key]);
  }
}

function item(value: unknown): void {
  const record = fields(value, ["objectId", "address", "size"]);
  uint64(record.objectId, true);
  const address = text(record.address, 18, 18);
  if (!/^0x[0-9a-f]{16}(?![\s\S])/.test(address)) invalid();
  uint64(record.size);
}

function items(value: unknown, max: number): void {
  for (const record of array(value, 0, max)) item(record);
}

function request(value: ObjectValue): void {
  fields(value, ["tag", "version", "requestId", "snapshotId", "operation", "args"]);
  equal(value.version, VERSION);
  uint64(value.requestId, true);
  switch (value.operation) {
    case "capabilities":
      equal(value.snapshotId, null);
      fields(value.args, []);
      break;
    case "snapshot.load": {
      equal(value.snapshotId, null);
      const args = fields(value.args, ["source", "delayMs"]);
      equal(args.source, "fixture:tiny");
      integer(args.delayMs, 0, 5000);
      break;
    }
    case "snapshot.dispose":
      snapshot(value.snapshotId);
      fields(value.args, []);
      break;
    case "query": {
      snapshot(value.snapshotId);
      const args = fields(value.args, ["text", "pageSize", "cursor"]);
      text(args.text, 1, 4096);
      integer(args.pageSize, 1, LIMITS.maxPageSize);
      cursor(args.cursor);
      break;
    }
    case "details": {
      snapshot(value.snapshotId);
      const args = fields(value.args, ["objectId", "pageSize", "cursor"]);
      uint64(args.objectId, true);
      integer(args.pageSize, 1, LIMITS.maxPageSize);
      cursor(args.cursor);
      break;
    }
    case "scene": {
      snapshot(value.snapshotId);
      const args = fields(value.args, ["maxItems"]);
      integer(args.maxItems, 1, LIMITS.maxSceneItems);
      break;
    }
    case "recipe.validate": {
      equal(value.snapshotId, null);
      const args = fields(value.args, ["recipe"]);
      const recipe = fields(args.recipe, ["schemaVersion", "query"]);
      equal(recipe.schemaVersion, VERSION);
      text(recipe.query, 1, 4096);
      break;
    }
    case "export": {
      snapshot(value.snapshotId);
      const args = fields(value.args, ["format", "maxBytes"]);
      equal(args.format, "svg");
      integer(args.maxBytes, 1, 32_768);
      break;
    }
    default:
      invalid();
  }
}

export function parseInbound(value: unknown): Inbound {
  const frame = object(value);
  switch (frame.tag) {
    case "hello": {
      fields(frame, ["tag", "versions", "extensions"]);
      const versions = array(frame.versions, 1, 8);
      for (const version of versions) {
        integer(version, 1, Number.MAX_SAFE_INTEGER);
      }
      if (new Set(versions).size !== versions.length) invalid();
      const extensions = array(frame.extensions, 0, 8);
      for (const extension of extensions) {
        const name = text(extension, 1, 64);
        if (!/^[a-z][a-z0-9]*(?:\.[a-z][a-z0-9]*)+(?![\s\S])/.test(name)) invalid();
      }
      if (new Set(extensions).size !== extensions.length) invalid();
      break;
    }
    case "request":
      request(frame);
      break;
    case "cancel":
      fields(frame, ["tag", "version", "requestId"]);
      equal(frame.version, VERSION);
      uint64(frame.requestId, true);
      break;
    case "shutdown":
      fields(frame, ["tag", "version"]);
      equal(frame.version, VERSION);
      break;
    default:
      invalid();
  }
  return frame as Inbound;
}

function result(value: unknown, snapshotId: unknown): void {
  const record = object(value);
  if (record.tag === "capabilities" || record.tag === "recipe") {
    equal(snapshotId, null);
  } else {
    snapshot(snapshotId);
  }
  switch (record.tag) {
    case "capabilities":
      fields(record, ["tag", "value"]);
      capabilities(record.value);
      break;
    case "snapshot":
      fields(record, ["tag", "objectCount"]);
      uint64(record.objectCount);
      break;
    case "disposed":
      fields(record, ["tag"]);
      break;
    case "page":
      fields(record, ["tag", "items", "nextCursor", "truncated"]);
      items(record.items, LIMITS.maxPageSize);
      cursor(record.nextCursor);
      boolean(record.truncated);
      break;
    case "scene":
      fields(record, ["tag", "schemaVersion", "items", "truncated"]);
      equal(record.schemaVersion, VERSION);
      items(record.items, LIMITS.maxSceneItems);
      boolean(record.truncated);
      break;
    case "recipe":
      fields(record, ["tag", "schemaVersion", "valid"]);
      equal(record.schemaVersion, VERSION);
      equal(record.valid, true);
      break;
    case "export":
      fields(record, ["tag", "format", "artifactId", "byteLength"]);
      equal(record.format, "svg");
      equal(record.artifactId, "fixture:svg");
      uint64(record.byteLength);
      break;
    default:
      invalid();
  }
}

export function parseOutbound(value: unknown): Outbound {
  const frame = object(value);
  switch (frame.tag) {
    case "ready":
      fields(frame, ["tag", "version", "capabilities"]);
      equal(frame.version, VERSION);
      capabilities(frame.capabilities);
      break;
    case "fatal":
      fields(frame, ["tag", "code", "message"]);
      member(frame.code, FATAL_CODES);
      text(frame.message, 1, 256);
      break;
    case "success":
      fields(frame, ["tag", "version", "requestId", "snapshotId", "result"]);
      equal(frame.version, VERSION);
      uint64(frame.requestId, true);
      result(frame.result, frame.snapshotId);
      break;
    case "error": {
      fields(frame, ["tag", "version", "requestId", "snapshotId", "error"]);
      equal(frame.version, VERSION);
      uint64(frame.requestId, true);
      nullableSnapshot(frame.snapshotId);
      const error = fields(frame.error, ["code", "message", "retryable"]);
      member(error.code, ERROR_CODES);
      text(error.message, 1, 256);
      boolean(error.retryable);
      break;
    }
    case "progress":
      fields(frame, ["tag", "version", "requestId", "snapshotId", "phase", "completed", "total"]);
      equal(frame.version, VERSION);
      uint64(frame.requestId, true);
      nullableSnapshot(frame.snapshotId);
      equal(frame.phase, "working");
      integer(frame.completed, 0, 100);
      equal(frame.total, 100);
      break;
    case "bye":
      fields(frame, ["tag", "version"]);
      equal(frame.version, VERSION);
      break;
    default:
      invalid();
  }
  return frame as Outbound;
}

File: src/MemoryVisualizer.Worker/Protocol.fs
namespace MemoryVisualizer.Worker

open System
open System.Collections.Generic
open System.Globalization
open System.Text
open System.Text.Json
open System.Text.RegularExpressions

exception ProtocolFailure of string

type Operation =
    | Capabilities
    | Load of delayMs: int
    | Dispose
    | Query of pageSize: int * cursor: uint64 option
    | Details of objectId: uint64 * pageSize: int * cursor: uint64 option
    | Scene of maxItems: int
    | ValidateRecipe
    | Export

type Request = {
    Id: uint64
    Snapshot: string option
    Operation: Operation
}

type Inbound =
    | Hello of versions: int64 array * extensions: string array
    | Request of Request
    | Cancel of uint64
    | Shutdown

[<RequireQualifiedAccess>]
module Protocol =
    [<Literal>]
    let MaxFrameBytes = 65536

    [<Literal>]
    let MaxOutstanding = 8

    let utf8 = UTF8Encoding(false, true)

    let operations = [|
        "capabilities"
        "snapshot.load"
        "snapshot.dispose"
        "query"
        "scene"
        "details"
        "recipe.validate"
        "export"
    |]

    let capabilities = {|
        backend = "fake"
        operations = operations
        extensions = Array.empty<string>
        limits = {|
            maxFrameBytes = MaxFrameBytes
            maxOutstanding = MaxOutstanding
            maxPageSize = 128
            maxSceneItems = 128
            progressIntervalMs = 100
        |}
    |}

    let private invalid () = raise (ProtocolFailure "InvalidFrame")

    let private require condition =
        if not condition then
            invalid ()

    let private fields (names: string list) (value: JsonElement) =
        require (value.ValueKind = JsonValueKind.Object)
        let actual = value.EnumerateObject() |> Seq.map _.Name |> Set.ofSeq
        require (actual = Set.ofList names)

    let private field (name: string) (value: JsonElement) = value.GetProperty(name)

    let private scalarLength (text: string) =
        let mutable index = 0
        let mutable count = 0

        while index < text.Length do
            if Char.IsHighSurrogate text[index] then
                require (index + 1 < text.Length && Char.IsLowSurrogate text[index + 1])
                index <- index + 2
            else
                require (not (Char.IsLowSurrogate text[index]))
                index <- index + 1

            count <- count + 1

        count

    let private text minLength maxLength (value: JsonElement) =
        require (value.ValueKind = JsonValueKind.String)
        let result = value.GetString()
        let length = scalarLength result
        require (length >= minLength && length <= maxLength)
        result

    let private literal expected value = require (text 0 4096 value = expected)

    let private boolValue (value: JsonElement) =
        require (value.ValueKind = JsonValueKind.True || value.ValueKind = JsonValueKind.False)

    let private digits (value: string) =
        value.Length > 0 && value |> Seq.forall (fun c -> c >= '0' && c <= '9')

    let private integer minimum maximum (value: JsonElement) =
        require (value.ValueKind = JsonValueKind.Number)
        let raw = value.GetRawText()
        require (digits raw)
        let mutable parsed = 0L
        require (Int64.TryParse(raw, NumberStyles.None, CultureInfo.InvariantCulture, &parsed))
        require (parsed >= minimum && parsed <= maximum)
        parsed

    let private decimalString minimum value =
        let raw = text 1 20 value
        require (digits raw && (raw.Length = 1 || raw[0] <> '0'))
        let mutable parsed = 0UL
        require (UInt64.TryParse(raw, NumberStyles.None, CultureInfo.InvariantCulture, &parsed))
        require (parsed >= minimum)
        parsed

    let private cursor (value: JsonElement) =
        if value.ValueKind = JsonValueKind.Null then
            None
        else
            Some(decimalString 0UL value)

    let private snapshot (value: JsonElement) =
        if value.ValueKind = JsonValueKind.Null then
            None
        else
            let raw = text 36 36 value
            let mutable parsed = Guid.Empty
            require (Guid.TryParseExact(raw, "D", &parsed))
            require (parsed <> Guid.Empty && raw = parsed.ToString("D"))
            Some raw

    let private array maximum (value: JsonElement) =
        require (value.ValueKind = JsonValueKind.Array && value.GetArrayLength() <= maximum)
        value.EnumerateArray() |> Seq.toArray

    let private version value = integer 1L 1L value |> ignore

    let private pageArgs value =
        let size = field "pageSize" value |> integer 1L 128L |> int
        let position = field "cursor" value |> cursor
        size, position

    let private parseRequest value =
        fields [ "tag"; "version"; "requestId"; "snapshotId"; "operation"; "args" ] value
        field "version" value |> version
        let id = field "requestId" value |> decimalString 1UL
        let scope = field "snapshotId" value |> snapshot
        let args = field "args" value

        let operation =
            match field "operation" value |> text 1 64 with
            | "capabilities" ->
                require scope.IsNone
                fields [] args
                Capabilities
            | "snapshot.load" ->
                require scope.IsNone
                fields [ "source"; "delayMs" ] args
                field "source" args |> literal "fixture:tiny"
                Load(field "delayMs" args |> integer 0L 5000L |> int)
            | "snapshot.dispose" ->
                require scope.IsSome
                fields [] args
                Dispose
            | "query" ->
                require scope.IsSome
                fields [ "text"; "pageSize"; "cursor" ] args
                field "text" args |> text 1 4096 |> ignore
                Query(pageArgs args)
            | "details" ->
                require scope.IsSome
                fields [ "objectId"; "pageSize"; "cursor" ] args
                let objectId = field "objectId" args |> decimalString 1UL
                let size, position = pageArgs args
                Details(objectId, size, position)
            | "scene" ->
                require scope.IsSome
                fields [ "maxItems" ] args
                Scene(field "maxItems" args |> integer 1L 128L |> int)
            | "recipe.validate" ->
                require scope.IsNone
                fields [ "recipe" ] args
                let recipe = field "recipe" args
                fields [ "schemaVersion"; "query" ] recipe
                field "schemaVersion" recipe |> version
                field "query" recipe |> text 1 4096 |> ignore
                ValidateRecipe
            | "export" ->
                require scope.IsSome
                fields [ "format"; "maxBytes" ] args
                field "format" args |> literal "svg"
                field "maxBytes" args |> integer 1L 32768L |> ignore
                Export
            | _ -> invalid ()

        {
            Id = id
            Snapshot = scope
            Operation = operation
        }

    let private parseInboundElement (value: JsonElement) =
        require (value.ValueKind = JsonValueKind.Object)

        match field "tag" value |> text 1 32 with
        | "hello" ->
            fields [ "tag"; "versions"; "extensions" ] value

            let versions =
                field "versions" value |> array 8 |> Array.map (integer 1L 9007199254740991L)

            require (versions.Length > 0 && (Array.distinct versions).Length = versions.Length)
            let extensions = field "extensions" value |> array 8 |> Array.map (text 1 64)
            require ((Array.distinct extensions).Length = extensions.Length)

            for extension in extensions do
                require (Regex.IsMatch(extension, "^[a-z][a-z0-9]*(\\.[a-z][a-z0-9]*)+\\z"))

            Hello(versions, extensions)
        | "request" -> Request(parseRequest value)
        | "cancel" ->
            fields [ "tag"; "version"; "requestId" ] value
            field "version" value |> version
            Cancel(field "requestId" value |> decimalString 1UL)
        | "shutdown" ->
            fields [ "tag"; "version" ] value
            field "version" value |> version
            Shutdown
        | _ -> invalid ()

    let private validateCapabilities value =
        fields [ "backend"; "operations"; "extensions"; "limits" ] value
        field "backend" value |> literal "fake"

        let actual =
            field "operations" value |> array operations.Length |> Array.map (text 1 64)

        require (actual = operations)
        field "extensions" value |> array 0 |> ignore
        let limits = field "limits" value

        fields
            [
                "maxFrameBytes"
                "maxOutstanding"
                "maxPageSize"
                "maxSceneItems"
                "progressIntervalMs"
            ]
            limits

        for name, expected in
            [
                "maxFrameBytes", 65536L
                "maxOutstanding", 8L
                "maxPageSize", 128L
                "maxSceneItems", 128L
                "progressIntervalMs", 100L
            ] do
            field name limits |> integer expected expected |> ignore

    let private validateItem value =
        fields [ "objectId"; "address"; "size" ] value
        field "objectId" value |> decimalString 1UL |> ignore
        let address = field "address" value |> text 18 18
        require (Regex.IsMatch(address, "^0x[0-9a-f]{16}\\z"))
        field "size" value |> decimalString 0UL |> ignore

    let private validateResult (value: JsonElement) =
        require (value.ValueKind = JsonValueKind.Object)

        match field "tag" value |> text 1 32 with
        | "capabilities" ->
            fields [ "tag"; "value" ] value
            field "value" value |> validateCapabilities
        | "snapshot" ->
            fields [ "tag"; "objectCount" ] value
            field "objectCount" value |> decimalString 0UL |> ignore
        | "disposed" -> fields [ "tag" ] value
        | "page" ->
            fields [ "tag"; "items"; "nextCursor"; "truncated" ] value
            field "items" value |> array 128 |> Array.iter validateItem
            field "nextCursor" value |> cursor |> ignore
            field "truncated" value |> boolValue
        | "scene" ->
            fields [ "tag"; "schemaVersion"; "items"; "truncated" ] value
            field "schemaVersion" value |> version
            field "items" value |> array 128 |> Array.iter validateItem
            field "truncated" value |> boolValue
        | "recipe" ->
            fields [ "tag"; "schemaVersion"; "valid" ] value
            field "schemaVersion" value |> version
            require ((field "valid" value).ValueKind = JsonValueKind.True)
        | "export" ->
            fields [ "tag"; "format"; "artifactId"; "byteLength" ] value
            field "format" value |> literal "svg"
            field "artifactId" value |> literal "fixture:svg"
            field "byteLength" value |> decimalString 0UL |> ignore
        | _ -> invalid ()

    let private validateOutboundElement (value: JsonElement) =
        require (value.ValueKind = JsonValueKind.Object)

        let correlation () =
            field "version" value |> version
            field "requestId" value |> decimalString 1UL |> ignore
            field "snapshotId" value |> snapshot |> ignore

        match field "tag" value |> text 1 32 with
        | "ready" ->
            fields [ "tag"; "version"; "capabilities" ] value
            field "version" value |> version
            field "capabilities" value |> validateCapabilities
        | "bye" ->
            fields [ "tag"; "version" ] value
            field "version" value |> version
        | "fatal" ->
            fields [ "tag"; "code"; "message" ] value
            let code = field "code" value |> text 1 64

            require (
                List.contains code [
                    "UnsupportedVersion"
                    "UnsupportedExtension"
                    "InvalidFrame"
                    "ProtocolViolation"
                    "TransportTimeout"
                    "InternalError"
                ]
            )

            field "message" value |> text 1 256 |> ignore
        | "progress" ->
            fields [ "tag"; "version"; "requestId"; "snapshotId"; "phase"; "completed"; "total" ] value
            correlation ()
            field "phase" value |> literal "working"
            field "completed" value |> integer 0L 100L |> ignore
            field "total" value |> integer 100L 100L |> ignore
        | "error" ->
            fields [ "tag"; "version"; "requestId"; "snapshotId"; "error" ] value
            correlation ()
            let error = field "error" value
            fields [ "code"; "message"; "retryable" ] error
            let code = field "code" error |> text 1 64

            require (
                List.contains code [
                    "Busy"
                    "Cancelled"
                    "SnapshotNotFound"
                    "InvalidRequest"
                    "NotImplemented"
                    "InternalError"
                    "StaleSnapshot"
                    "WorkerExited"
                    "Timeout"
                    "ProtocolError"
                ]
            )

            field "message" error |> text 1 256 |> ignore
            field "retryable" error |> boolValue
        | "success" ->
            fields [ "tag"; "version"; "requestId"; "snapshotId"; "result" ] value
            correlation ()
            let result = field "result" value
            validateResult result
            let scope = field "snapshotId" value |> snapshot

            match field "tag" result |> text 1 32 with
            | "capabilities"
            | "recipe" -> require scope.IsNone
            | _ -> require scope.IsSome
        | _ -> invalid ()

    let private withDocument validator (bytes: byte array) =
        try
            require (bytes.Length > 0 && bytes.Length <= MaxFrameBytes)
            require (not (bytes |> Array.exists (fun b -> b = 10uy || b = 13uy)))
            utf8.GetString(bytes) |> ignore

            use document =
                JsonDocument.Parse(ReadOnlyMemory<byte>(bytes), JsonDocumentOptions(MaxDepth = 16))

            require (document.RootElement.ValueKind = JsonValueKind.Object)

            let rec validateTree (value: JsonElement) =
                match value.ValueKind with
                | JsonValueKind.Object ->
                    let names = HashSet<string>(StringComparer.Ordinal)

                    for property in value.EnumerateObject() do
                        scalarLength property.Name |> ignore
                        require (names.Add property.Name)
                        validateTree property.Value
                | JsonValueKind.Array ->
                    for child in value.EnumerateArray() do
                        validateTree child
                | JsonValueKind.String -> value.GetString() |> scalarLength |> ignore
                | JsonValueKind.Number -> integer 0L 9007199254740991L value |> ignore
                | _ -> ()

            validateTree document.RootElement
            validator document.RootElement
        with
        | :? JsonException
        | :? DecoderFallbackException
        | :? InvalidOperationException
        | :? ArgumentException
        | :? KeyNotFoundException -> invalid ()

    let validateJson bytes = withDocument ignore bytes

    let parseInbound bytes = withDocument parseInboundElement bytes

    let validateOutbound bytes =
        withDocument validateOutboundElement bytes

    let private encode (value: obj) =
        let bytes = JsonSerializer.SerializeToUtf8Bytes(value, value.GetType())
        validateOutbound bytes
        bytes

    let scopeValue (scope: string option) = Option.toObj scope

    let idValue (id: uint64) =
        id.ToString(CultureInfo.InvariantCulture)

    let ready () =
        encode {|
            tag = "ready"
            version = 1
            capabilities = capabilities
        |}

    let bye () = encode {| tag = "bye"; version = 1 |}

    let message =
        function
        | "UnsupportedVersion" -> "No supported protocol version."
        | "UnsupportedExtension" -> "Requested extension is not supported."
        | "InvalidFrame" -> "Invalid protocol frame."
        | "ProtocolViolation" -> "Protocol sequence violation."
        | "TransportTimeout" -> "Transport deadline exceeded."
        | "Busy" -> "Request limit reached."
        | "Cancelled" -> "Request cancelled."
        | "SnapshotNotFound" -> "Snapshot is not active."
        | "InvalidRequest" -> "Invalid request."
        | _ -> "Internal worker error."

    let fatal code =
        encode {|
            tag = "fatal"
            code = code
            message = message code
        |}

    let error (request: Request) code =
        encode {|
            tag = "error"
            version = 1
            requestId = idValue request.Id
            snapshotId = scopeValue request.Snapshot
            error = {|
                code = code
                message = message code
                retryable = code = "Busy"
            |}
        |}

    let success (request: Request) (scope: string option) (result: obj) =
        encode {|
            tag = "success"
            version = 1
            requestId = idValue request.Id
            snapshotId = scopeValue scope
            result = result
        |}

    let progress (request: Request) (completed: int) =
        encode {|
            tag = "progress"
            version = 1
            requestId = idValue request.Id
            snapshotId = scopeValue request.Snapshot
            phase = "working"
            completed = completed
            total = 100
        |}
